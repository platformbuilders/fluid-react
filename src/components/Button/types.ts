import { ReactNode } from 'react';
import {
  ButtonColorType,
  ButtonVariants,
  TypographyVariants,
} from '@platformbuilders/theme-toolkit';
import RadixIcon from '@radix-ui/react-icons';

export type ButtonProps = {
  style?: any;
  type?: 'button' | 'submit' | 'reset';
  loading?: boolean;
  $fullWidth?: boolean;
  variant?: ButtonVariants;
  typographyVariant?: TypographyVariants;
  children?: string | ReactNode;
  leftIconName?: keyof typeof RadixIcon;
  rightIconName?: keyof typeof RadixIcon;
  colorVariant?: ButtonColorType;
  size?: 'normal' | 'medium' | 'large';
} & TouchableType;

export type TouchableType = {
  id?: string;
  accessibility: string;
  testID?: string;
  disabled?: boolean;
  onPress?: (param: any) => void;
};
