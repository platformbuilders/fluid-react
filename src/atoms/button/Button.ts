import styled from 'styled-components';

export const Button = styled.button`
  height: 48px;
  padding: 0 16px;
  background-color: #305bc4;
  border: none;
  border-radius: 4px;
  color: #ffffff;
  font-size: 15px;
  font-weight: 700;
  text-transform: uppercase;
  letter-spacing: 0.05em;
`;
