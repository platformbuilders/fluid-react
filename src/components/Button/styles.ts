import styled from 'styled-components/native';
import TouchableComponent from '../Touchable';
import { getTheme } from '../../utils/helpers';

const primaryMain = getTheme('primary.main');
const secondaryMain = getTheme('secondary.main');
const primaryContrast = getTheme('primary.contrast');
const primaryDark = getTheme('primary.dark');

interface TouchableProps {
  rounded: boolean;
}
export const Touchable = styled(TouchableComponent)<TouchableProps>`
  border-radius: ${(props): string => (props.rounded ? '50px' : '0')};
`;

interface ButtonWrapperProps {
  rounded: boolean;
  terciary: boolean;
  secondary: boolean;
  disabled?: boolean;
}
export const ButtonWrapper = styled.View<ButtonWrapperProps>`
  height: 42px;
  flex-direction: row;
  border: 2px solid;
  align-items: center;
  margin-vertical: 6px;
  min-width: 180px;
  padding: ${(props): string => (props.rounded ? '0' : '10px 11px')};
  border-radius: ${(props): string => (props.rounded ? '50px' : '24px')}
  justify-content: center;
  background-color: ${(props): string =>
    props.terciary
      ? 'transparent'
      : props.secondary
      ? secondaryMain(props)
      : props.disabled
      ? primaryDark(props)
      : primaryMain(props)};
  border-color: ${(props): string =>
    props.terciary
      ? 'transparent'
      : props.secondary
      ? secondaryMain(props)
      : primaryDark(props)};
`;

interface TextButtonProps {
  terciary: boolean;
  secondary: boolean;
  disabled?: boolean;
}
export const TextButton = styled.Text<TextButtonProps>`
  color: ${(props): string =>
    props.secondary
      ? primaryMain(props)
      : props.terciary
      ? primaryDark(props)
      : primaryContrast(props)};
`;
