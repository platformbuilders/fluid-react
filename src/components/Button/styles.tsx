import styled, { css } from 'styled-components';
import {
  ThemeProps,
  getTheme,
  ifStyle,
  pxToRem,
} from '@platformbuilders/theme-toolkit';
import LoadingIndicator from '../LoadingIndicator';
import TouchableComponent from '../Touchable';
import Typography from '../Typography';
import { ButtonProps } from './types';

const buttonRadius = getTheme('themeRadius.button');
const spacingSm = getTheme('spacing.sm');
const borderWidthSm = getTheme('borderWidth.sm');
const isDisabled = ifStyle('disabled');

type ButtonWrapperProps = ButtonProps & ThemeProps;

const getStylesButton = (props: ButtonWrapperProps) => {
  const colorButton =
    getTheme(`brand.${props.colorVariant}.main`)(props) ||
    getTheme(`${props.colorVariant}.main`)(props);
  const colorConstrastButton =
    getTheme(`brand.${props.colorVariant}.contrast`)(props) ||
    getTheme(`${props.colorVariant}.contrast`)(props);
  const defaultTheme = css`
    background-color: ${colorButton};
    box-shadow: 0px 1px 3px 0px rgba(0, 0, 0, 0.1);
    &:hover {
      background-color: ${colorButton}90;
    }
    .text-button {
      color: ${colorConstrastButton};
    }
    svg {
      color: ${colorConstrastButton};
    }
  `;
  switch (props.variant) {
    case 'filled':
      return defaultTheme;
    case 'ghost':
      return css`
        background-color: transparent;
        &:hover {
          background-color: ${colorButton}10;
        }
        .text-button {
          color: ${colorButton};
        }
        svg {
          color: ${colorButton};
        }
      `;

    case 'tint':
      return css`
        background-color: ${colorButton}10;
        &:hover {
          background-color: ${colorButton}20;
        }
        .text-button {
          color: ${colorButton};
        }
        svg {
          color: ${colorButton};
        }
      `;

    case 'outline':
      return css`
        background-color: transparent;
        border: ${borderWidthSm}px solid ${colorButton};
        &:hover {
          background-color: ${colorButton}10;
        }
        .text-button {
          color: ${colorButton};
        }
        svg {
          color: ${colorButton};
        }
      `;

    default:
      return defaultTheme;
  }
};

const sizeButton = {
  normal: css`
    height: 2.25rem;
  `,
  medium: css`
    height: 2.75rem;
  `,
  large: css`
    height: 3.25rem;
  `,
};

export const Touchable = styled(TouchableComponent)<ButtonWrapperProps>`
  border: none;
  ${getStylesButton}
  ${({ size }) => !!size && sizeButton[size]}
  width: ${({ $fullWidth }) => (!!$fullWidth ? '100%' : undefined)};
  padding: ${spacingSm}px;
  cursor: ${isDisabled('not-allowed', 'pointer')};
  flex-direction: row;
  align-items: center;
  overflow: hidden;
  border-radius: ${buttonRadius}px;
  justify-content: center;
  transition: all 0.2s ease-in-out;
  position: relative;
  display: inline-flex;
  box-sizing: border-box;
  vertical-align: middle;
  text-align: center;
  text-overflow: ellipsis;
  outline: none;

  &::-moz-focus-inner {
    border: none;
  }
  &::before {
    content: '';
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    background-color: rgb(255, 255, 255);
    opacity: 0;
    transition: opacity 0.2s;
  }
  &::after {
    content: '';
    position: absolute;
    left: 50%;
    top: 50%;
    border-radius: 50%;
    padding: 50%;
    width: 32px; /* Safari */
    height: 32px; /* Safari */
    background-color: rgb(255, 255, 255);
    opacity: 0;
    transform: translate(-50%, -50%) scale(1);
    transition: opacity 1s, transform 0.5s;
  }
  &:active {
    box-shadow: 0 5px 5px -3px rgba(0, 0, 0, 0.2),
      0 8px 10px 1px rgba(0, 0, 0, 0.14), 0 3px 14px 2px rgba(0, 0, 0, 0.12);
  }
  &:active::after {
    opacity: 0.3;
    transform: translate(-50%, -50%) scale(0);
    transition: transform 0s;
  }
  &:disabled {
    opacity: 0.3;
    cursor: initial;
  }
  &:disabled::before {
    opacity: 0;
  }
  &:disabled::after {
    opacity: 0;
  }
`;

export const ContentWrapper = styled.div<ButtonWrapperProps>`
  display: flex;
  align-items: center;
  gap: 8px;
  flex-direction: row;

  svg {
    width: 20px;
    height: 20px;
  }
`;

export const TextButton = styled(Typography)`
  letter-spacing: 0.4px;
  font-style: normal;
  font-weight: 600;
  line-height: 100%;
  letter-spacing: 0.8px;
`;

export const Loading = styled(LoadingIndicator).attrs({
  variant: 'button',
})`
  align-self: center;
  width: ${pxToRem(55)}px;
`;
