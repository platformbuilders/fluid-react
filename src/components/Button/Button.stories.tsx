import type { Meta, StoryObj } from '@storybook/react';
import Button from './index';

const meta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button,

  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  args: {
    children: 'Button',
  },
};

export default meta;

type Story = StoryObj<typeof Button>;

export const Normal: Story = {
  args: {
    children: 'Button',
    size: 'normal',
  },
};

export const Medium: Story = {
  args: {
    ...Normal.args,
    size: 'medium',
  },
};

export const Large: Story = {
  args: {
    ...Normal.args,
    size: 'large',
  },
};
