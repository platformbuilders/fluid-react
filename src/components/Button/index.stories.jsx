import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text, boolean } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import markdown from './button.md';
import Button from './index';

storiesOf('Button', module)
  .addDecorator(withKnobs)
  .addParameters({
    readme: {
      content: markdown,
    },
    viewport: {
      defaultViewport: '',
    },
  })
  .add('Example', () => (
    <Button
      onPress={action('clicked')}
      loading={boolean('loading', false)}
      secondary={boolean('secondary', false)}
      transparent={boolean('transparent', false)}
      disabled={boolean('disabled', false)}
    >
      {text('Text', 'Button Test')}
    </Button>
  ));
