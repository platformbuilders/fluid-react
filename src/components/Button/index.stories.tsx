import { action } from '@storybook/addon-actions';
import type { Meta, StoryObj } from '@storybook/react';
// import { boolean, text, withKnobs } from '@storybook/addon-knobs';
// import { storiesOf } from '@storybook/react';
// import markdown from './button.md';
import Button from './index';

// storiesOf('Button', module)
//   .addDecorator(withKnobs)
//   .addParameters({
//     readme: {
//       content: markdown,
//     },
//     viewport: {
//       defaultViewport: '',
//     },
//   })
//   .add('Example', () => (
//     <Button
//       onPress={action('clicked')}
//       loading={boolean('loading', false)}
//       transparent={boolean('transparent', false)}
//       disabled={boolean('disabled', false)}
//     >
//       {text('Text', 'Button Test')}
//     </Button>
//   ));

const meta: Meta<typeof Button> = {
  title: 'Button',
  component: Button,
  argTypes: {
    loading: { control: 'boolean' },
    disabled: { control: 'boolean' },
    id: { control: 'text' },
    testID: { control: 'text' },
  },
  args: {
    children: 'Button test',
    accessibility: 'button',
    onPress: action('clicked'),
    loading: false,
  },
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Default: Story = {};
