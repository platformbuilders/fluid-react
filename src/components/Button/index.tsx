import React, { FC, useContext } from 'react';
import { ThemeProvider } from 'styled-components';
import LoadingIndicator from '../LoadingIndicator';
import { ThemeContext } from '../ThemeContext';
import { Touchable, ButtonWrapper, TextButton } from './styles';

type Props = {
  children: string;
  onPress(): void;
  id: string;
  accessibility: string;
  style?: object[];
  textStyle?: object;
  disabled?: boolean;
  rounded?: boolean;
  secondary?: boolean;
  terciary?: boolean;
  loading?: boolean;
};

const Button: FC<Props> = ({
  children,
  onPress,
  accessibility,
  style = [{}],
  textStyle = {},
  disabled = false,
  rounded = false,
  secondary = false,
  terciary = false,
  loading = false,
}) => {
  const { theme } = useContext(ThemeContext);

  return (
    <ThemeProvider theme={theme}>
      <Touchable
        accessibility={accessibility}
        disabled={loading || disabled}
        onPress={onPress}
        rounded={rounded}
      >
        <ButtonWrapper
          secondary={secondary}
          terciary={terciary}
          style={style}
          disabled={disabled}
          rounded={rounded}
        >
          {loading && <LoadingIndicator />}
          {!loading && (
            <>
              <TextButton
                secondary={secondary}
                terciary={terciary}
                style={textStyle}
                disabled={disabled}
              >
                {children}
              </TextButton>
            </>
          )}
        </ButtonWrapper>
      </Touchable>
    </ThemeProvider>
  );
};

export default Button;
