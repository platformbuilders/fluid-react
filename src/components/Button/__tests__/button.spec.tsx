import { shallow } from 'enzyme';
import { ThemeProvider } from 'styled-components';

import Button from '..';
import theme from '../../../theme';
import { ButtonProps as Props } from '../../../types';

const defaultContent = 'Text';
const defaultProps: Props = {
  accessibility: 'label',
  onPress: jest.fn(),
  children: defaultContent,
};

describe('Component: Button', () => {
  test('snapshots with default props', () => {
    const component = shallow(
      <ThemeProvider theme={theme}>
        <Button {...defaultProps} />
      </ThemeProvider>,
    );

    expect(component).toMatchSnapshot();
  });

  test('snapshots with loading', () => {
    const component = shallow(
      <ThemeProvider theme={theme}>
        <Button {...defaultProps} loading />
      </ThemeProvider>,
    );

    expect(component).toMatchSnapshot();
  });

  test('snapshots with other props', () => {
    const component = shallow(
      <ThemeProvider theme={theme}>
        <Button {...defaultProps} disabled variant="secondary">
          {defaultContent}
        </Button>
      </ThemeProvider>,
    );

    expect(component).toMatchSnapshot();
  });

  test('check props default props', () => {
    // when
    const component = shallow(
      <Button {...defaultProps} disabled variant="secondary">
        {defaultContent}
      </Button>,
    );

    // then
    const child = component.children();

    expect(child.contains(defaultContent)).toEqual(true);
  });

  test('should call onPress when pressed', () => {
    // should
    const mockFunction = jest.fn();
    const component = shallow(
      <Button {...defaultProps} onPress={mockFunction}>
        {defaultContent}
      </Button>,
    );

    // when
    expect(mockFunction).not.toHaveBeenCalled();
    component.props().onPress();

    // then
    expect(mockFunction).toHaveBeenCalled();
  });
  test('should not call onPress when pressed', () => {
    // should
    const mockFunction = jest.fn();
    const component = shallow(
      <Button {...defaultProps} disabled onPress={mockFunction}>
        {defaultContent}
      </Button>,
    );

    component.simulate('click');
    expect(mockFunction).not.toHaveBeenCalled();
  });
});
