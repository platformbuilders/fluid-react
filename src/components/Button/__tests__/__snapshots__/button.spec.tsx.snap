// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: Button snapshots with default props 1`] = `
<div>
  <button
    accessibility="label"
    class="sc-bcHOTY sc-dkHyXG fVJZLw"
    id="label"
    variant="primary"
  >
    <p
      class="sc-gsMHZj bWuTyU sc-hLljAB cvQGLR"
      variant="md"
    >
      Text
    </p>
  </button>
</div>
`;

exports[`Component: Button snapshots with loading 1`] = `
<div>
  <button
    accessibility="label"
    class="sc-bcHOTY sc-dkHyXG knOvRL"
    disabled=""
    id="label"
    variant="primary"
  />
</div>
`;

exports[`Component: Button snapshots with other props 1`] = `
<div>
  <button
    accessibility="label"
    class="sc-bcHOTY sc-dkHyXG knOvRL"
    disabled=""
    id="label"
    variant="secondary"
  >
    <p
      class="sc-gsMHZj bWuTyU sc-hLljAB cvQGLR"
      disabled=""
      variant="md"
    >
      Text
    </p>
  </button>
</div>
`;
